File: src/dp/lc96.cc
#include <vector>

using namespace std;

class Solution {
 public:
  int numTrees(int n) {
    auto f = vector<int>(n + 1, 0);
    f[0] = 1;
    f[1] = 1;
    for (auto i = 2; i <= f.size(); i++) {
      for (auto j = 1; j <= i; j++) {
        f[i] += f[j - 1] * f[i - j];
      }
    }
    return f[n];
  }
};

File: src/dp/lc96.js
/**
 * @param {number} n
 * @return {number}
 */
function numTrees(n) {
  const f = Array.from({ length: n + 1 }, () => 0);
  f[0] = 1;
  f[1] = 1;
  for (let i = 2; i <= n; i++) {
    for (let j = 0; j < i; j++) {
      f[i] += f[j] * f[i - 1 - j];
    }
  }

  return f[n];
}

export default numTrees;
